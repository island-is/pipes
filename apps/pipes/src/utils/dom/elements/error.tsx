/* eslint-disable @typescript-eslint/no-use-before-define */
import React from "react";

import { Dialog } from "./dialog.js";

import type { ReactNode } from "react";

export interface IError {
  type: "Error";
  title?: string | undefined | null;
  children?: ReactNode | ReactNode[];
}
export const _Error = (props: Omit<IError, "type">): ReactNode => {
  const children = (Array.isArray(props.children) ? [...props.children] : [props.children]).filter(Boolean);
  return renderError.ansi({
    type: "Error",
    ...props,
    children,
  });
};

export const renderError = {
  ansi: (component: IError): ReactNode => {
    return (
      <Dialog title={component.title ?? "Error"} dialogType={"error"}>
        {component.children}
      </Dialog>
    );
  },

  markdown: (_component: IError): ReactNode => {
    throw "Not implemented";
  },
};
