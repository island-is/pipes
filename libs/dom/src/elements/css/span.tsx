import { applyANSIStyle, escapeAnsi } from "./ansi-escape.js";
import { addBorder } from "./border.js";
import { CSS_COLORS } from "./colors.js";
import { computeCSS } from "./css.js";
import { splitByHeight, splitByWidth } from "./split-helpers.js";

import type { CSS } from "./types.js";

export const span = (text: string, _css: CSS, width: number = process.stdout.columns || 80): string[] => {
  const css = computeCSS(_css, width);
  const borderHeight = (css.borderTop ? 1 : 0) + (css.borderBottom ? 1 : 0);
  const marginHeight = (css.marginTop ? css.marginTop : 0) + (css.marginBottom ? css.marginBottom : 0);
  const paddingHeight = (css.paddingTop ? css.paddingTop : 0) + (css.paddingBottom ? css.paddingBottom : 0);
  const outsideHeight = borderHeight + marginHeight + paddingHeight;
  const marginWidth = (css.marginLeft ?? 0) + (css.marginRight ?? 0);
  const borderWidth = (css.borderLeft ? 1 : 0) + (css.borderRight ? 1 : 0);
  const paddingLeft = css.paddingLeft ?? 0;
  const expectedWidth = css.width ? css.width - marginWidth - borderWidth - 1 : "auto";
  let expectedHeight = typeof css.height !== "undefined" ? Math.max(css.height - outsideHeight, 0) : ("auto" as const);
  let marginTopStrArr = Array(css.marginTop ?? 0).fill("");
  let marginBottomStrArr = Array(css.marginBottom ?? 0).fill("");
  const paddingTopStrArr = Array(css.paddingTop ?? 0).fill("");
  const paddingBottomStrArr = Array(css.paddingBottom ?? 0).fill("");
  if (expectedHeight !== "auto") {
    expectedHeight = Math.max(expectedHeight - outsideHeight, 0);
  }
  if (expectedHeight === 0 || css.hidden) {
    return [];
  }
  let lines = text.split("\n").flatMap((line) => {
    return splitByWidth(line, expectedWidth, css.paddingLeft ?? 0, css.paddingRight ?? 0, css.textAlign);
  });

  if (typeof expectedHeight === "number" && lines.length > expectedHeight) {
    lines = splitByHeight(lines, expectedHeight);
  }

  if (expectedWidth !== "auto") {
    marginBottomStrArr = marginBottomStrArr.map((e) => e.padEnd(expectedWidth));
    marginTopStrArr = marginTopStrArr.map((e) => e.padEnd(expectedWidth));
  }
  lines = [...paddingTopStrArr, ...lines, ...paddingBottomStrArr];
  const lineLength = lines.reduce((a, b) => Math.max(a, escapeAnsi(b).length), 0);

  // Apply font styles
  if (css.visibility === true && css.fontStyle) {
    for (const style of css.fontStyle) {
      for (let i = 0; i < lines.length; i++) {
        lines[i] = applyANSIStyle(lines[i].padEnd(lineLength), style.ansi);
      }
    }
  }

  if (css.visibility === true && css.color) {
    for (let i = 0; i < lines.length; i++) {
      lines[i] = applyANSIStyle(lines[i].padEnd(lineLength), {
        startTag: css.color.ansi.fg,
        endTag: CSS_COLORS.DEFAULT.ansi.fg,
      });
    }
  }

  if (css.visibility === true && css.backgroundColor) {
    for (let i = 0; i < lines.length; i++) {
      lines[i] = applyANSIStyle(lines[i], {
        startTag: css.backgroundColor.ansi.bg,
        endTag: CSS_COLORS.DEFAULT.ansi.bg,
      });
    }
  }
  const realBorderWidth =
    typeof expectedWidth === "number" ? expectedWidth : lines.reduce((a, b) => Math.max(escapeAnsi(b).length, a), 0);
  lines = [...marginTopStrArr, ...addBorder(lines, css, realBorderWidth), ...marginBottomStrArr];
  if (!css.visibility) {
    return lines.map((e) => e.split("").fill(" ").join(""));
  }
  return lines;
};
