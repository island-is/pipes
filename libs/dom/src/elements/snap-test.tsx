import assert from "node:assert";
import { basename, dirname, join } from "node:path";
import { fileURLToPath } from "node:url";

import { SnapshotState } from "jest-snapshot";

// eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
function toMatchSnapshot(actual: any, testFile: string, testTitle: string) {
  const root = join(fileURLToPath(dirname(testFile)), "__snapshots__");
  const fileName = `${basename(fileURLToPath(testFile))}.snap`;
  const snapshotFile = join(root, fileName);
  const snapshotState = new SnapshotState(snapshotFile, {
    updateSnapshot: process.env.SNAPSHOT_UPDATE ? "all" : "new",
  } as any);

  const matcher = snapshotState.match({
    testName: `${testTitle}-${testFile}`,
    received: actual,
    isInline: false,
  });

  const result = matcher;
  snapshotState.save();

  return result;
}

// eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
export const snapTest = (actual: any, testFile: string, testTitle: string) => {
  const result = toMatchSnapshot(actual, testFile, testTitle);
  assert(result.pass, `${testTitle} - does not match snapshot`);
};
