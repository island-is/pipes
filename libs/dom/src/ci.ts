/* eslint-disable @typescript-eslint/no-use-before-define */
import ciinfo from "ci-info";
import stripAnsi from "strip-ansi";

import { getDifferences } from "./compare.js";
import { renderContainer } from "./elements/container.js";
import { span } from "./elements/css/span.js";
import { renderBadge } from "./elements/elements.js";
import { renderError } from "./elements/error.js";
import { renderFailure } from "./elements/failure.js";
import { renderGroup } from "./elements/group.js";
import { renderLink } from "./elements/link.js";
import { renderMask } from "./elements/mask.js";
import { renderRow } from "./elements/row.js";
import { renderSuccess } from "./elements/success.js";
import { renderText } from "./elements/text.js";
import { renderTimestamp } from "./elements/timestamp.js";
import { isListItem } from "./helpers.js";

import type * as elementsJs from "./elements/elements.js";

const ComponentRenderersANSI: { [key: string]: (component: any, width: number) => string } = {
  Mask: renderMask.ansi,
  Group: renderGroup.ansi(renderToANSIString),
  Text: renderText.ansi(renderToANSIString),
  Link: renderLink.ansi(renderToANSIString),
  Timestamp: renderTimestamp.ansi(renderToANSIString),
  Container: renderContainer.ansi(renderToANSIString),
  Row: renderRow.ansi(renderToANSIString),
  Success: renderSuccess.ansi(renderToANSIString),
  Failure: renderFailure.ansi(renderToANSIString),
  Error: renderError.ansi(renderToANSIString),
  Info: (component: elementsJs.IInfo, width: number) => {
    return span(`${renderToANSIString(component.children as any, width)}`, {
      width,
      color: "cyan",
    }).join("");
  },
  Log: (component: elementsJs.ILog, width: number) => {
    return span(`${renderToANSIString(component.children as any, width)}`, {
      width,
    }).join("");
  },
  Title: (component: elementsJs.ITitle, width: number) => {
    return span(`${renderToANSIString(component.children as any, width)}`, {
      width,
      textAlign: "center",
      margin: {
        bottom: 0,
      },

      padding: {
        top: 2,
        bottom: 1,
      },
      fontStyle: "bold",
      border: {
        bottom: true,
      },
    }).join("");
  },
  Subtitle: (component: elementsJs.ISubtitle, width: number) => {
    return span(`${renderToANSIString(component.children as any, width)}`, {
      width,
      fontStyle: "underline",
    }).join("");
  },
  List: (component: elementsJs.IList, width: number) => {
    if (Array.isArray(component.children)) {
      const children = component.children.filter((e): e is elementsJs.IListItem => isListItem(e));
      return children
        .map((child) =>
          span(`- ${renderToANSIString(child, width)}`, {
            width,
            margin: {
              left: 2,
            },
          }),
        )
        .join("\n");
    } else {
      if (isListItem(component.children)) {
        return span(`- ${renderToANSIString(component.children, width - 2)}`, {
          width,
          margin: {
            left: 2,
          },
        }).join("");
      }
    }
    return "";
  },
  ListItem: (component: elementsJs.IListItem, width: number) => {
    return span(renderToANSIString(component.children as any, width), {
      width,
    }).join("");
  },
  Dialog: (component: elementsJs.IDialog, width: number) => {
    const color =
      component.dialogType === "error"
        ? ("white" as const)
        : component.dialogType === "failure"
        ? ("white" as const)
        : component.dialogType === "success"
        ? ("white" as const)
        : ("white" as const);
    const backgroundColor =
      component.dialogType === "error"
        ? ("red" as const)
        : component.dialogType === "failure"
        ? ("red" as const)
        : component.dialogType === "success"
        ? ("green" as const)
        : ("black" as const);
    return span(component.message, {
      width,
      color,
      backgroundColor,
      border: {
        top: true,
        bottom: true,
        left: true,
        right: true,
        style: "solid",
      },

      padding: {
        left: 2,
        top: component.paddingTop,
        bottom: component.paddingBottom,
      },
    }).join("");
  },
  Badge: renderBadge.ansi(renderToANSIString),
  Divider: (component: elementsJs.IDivider, width: number) => {
    return span(" ", {
      width,
      height: 2,
      border: {
        bottom: true,
        style: "solid",
      },
    }).join("");
  },
  Note: (component: elementsJs.INote, width: number) => {
    return span(`Note: ${renderToANSIString(component.children as any, width)}`, {
      width,
      color: "cyan",
    }).join("");
  },
  Warning: (component: elementsJs.IWarning, width: number) => {
    return span(`WARNING: ${renderToANSIString(component.children as any, width)}`, {
      width,
      color: "yellow",
    }).join("");
  },
  Highlight: (component: elementsJs.IHighlight, width: number) => {
    return span(`${renderToANSIString(component.children as any, width)}`, {
      width,
      color: "black",
      backgroundColor: "white",
    }).join("");
  },
  Fragment: (component: elementsJs.IFragment, width: number) => {
    if (Array.isArray(component.children)) {
      return component.children.map((child) => renderToANSIString(child as any, width)).join("");
    } else {
      return renderToANSIString(component.children, width);
    }
  },
  Code: (component: elementsJs.ICode) => {
    // TODO: Fix this
    return component.children?.toString() ?? "";
  },
  Table: (component: elementsJs.ITable, width: number) => {
    const children = component.children
      ? Array.isArray(component.children)
        ? component.children
        : [component.children]
      : [];
    return children
      .map((e) => renderToANSIString(e as elementsJs.PipeComponents, width))
      .flat()
      .join("");
  },
  TableRow: (component: elementsJs.ITableRow, width: number) => {
    const children = component.children
      ? Array.isArray(component.children)
        ? component.children
        : [component.children]
      : [];
    const count = children.length;
    const childWidth = Math.floor(width / count);
    return children
      .map((e) =>
        span(renderToANSIString(e as elementsJs.PipeComponents, childWidth - 2), {
          width: childWidth,
          padding: {
            left: 1,
            right: 1,
          },
        }),
      )
      .join("");
  },
  TableCell: (component: elementsJs.ITableCell, width: number) => {
    return renderToANSIString(component.children, width);
  },
  TableHeadings: (component: elementsJs.ITableHeadings, width: number) => {
    const children = component.children
      ? Array.isArray(component.children)
        ? component.children
        : [component.children]
      : [];
    const count = children.length;
    const childWidth = Math.floor(width / count);
    return `${children
      .map((e) =>
        span(renderToANSIString(e as elementsJs.PipeComponents, childWidth), {
          width: childWidth,
          fontStyle: "bold",
          padding: {
            left: 0,
            right: 0,
          },
        }),
      )
      .join("")}`;
  },
};

export function renderToANSIString(component: elementsJs.PipeComponents | null, width: number): string {
  if (component === null || typeof component === "undefined") {
    return "";
  }
  if (typeof component === "string" || typeof component == "number") {
    return `${component}`.trimStart();
  }
  if (Array.isArray(component)) {
    return component
      .flat(10)
      .map((value) => renderToANSIString(value, width))
      .join(" ");
  }

  if (!component || !("type" in component)) {
    return "";
  }

  if (ComponentRenderersANSI[component.type as string]) {
    return ComponentRenderersANSI[component.type as string](component, width);
  }
  throw new Error("Invalid type");
}

class ConsoleRender {
  #PRINT_ONLY_CHANGES = false;
  #TERMINAL_WIDTH: number;
  #elements: elementsJs.PipeComponents[] = [];
  #renderInProgress = false;
  constructor() {
    if (ciinfo.isCI || typeof process.stdout.columns == "undefined") {
      this.#TERMINAL_WIDTH = 120;
    } else {
      this.#TERMINAL_WIDTH = 60;
      process.stdout.on("resize", () => {
        this.#TERMINAL_WIDTH = process.stdout.columns;
      });
    }
  }
  #lockRender = async (fn: () => void | Promise<void>) => {
    await this.#waitForRenderLock();
    this.#renderInProgress = true;
    await fn();
    this.#renderInProgress = false;
  };
  #waitForRenderLock = () => {
    return new Promise<void>((resolve) => {
      const fn = () => {
        if (!this.#renderInProgress) {
          resolve();
          return;
        }
        setTimeout(fn, 50);
      };
      fn();
    });
  };
  #streamWrite = (chunk: string, stream = process.stdout, encoding: BufferEncoding = "utf-8") => {
    return new Promise<void>((resolve, reject) => {
      const errListener = (err: any) => {
        stream.removeListener("error", errListener);
        reject(err);
      };
      stream.addListener("error", errListener);
      const callback = () => {
        stream.removeListener("error", errListener);
        resolve();
      };
      stream.write(chunk, encoding, callback);
    });
  };

  async mountAndRender(element: elementsJs.PipeComponents) {
    this.#elements.push(element);
    const indexOf = this.#elements.indexOf(element);

    if (this.#PRINT_ONLY_CHANGES) {
      await this.#render(element);
    } else {
      await this.#render(element);
    }
    const render = async (newElements: elementsJs.PipeComponents) => {
      if (this.#PRINT_ONLY_CHANGES) {
        const prevElements = this.#elements[indexOf];
        const compare = getDifferences(newElements, prevElements);
        this.#elements[indexOf] = newElements;
        await this.#render(compare as any);

        return;
      }
      this.#elements[indexOf] = newElements;
      await this.#render(newElements);
    };
    return render;
  }
  #splitByNewlineOrLength(inputString: string) {
    const initialLines = inputString.split("\n");
    const resultLines = [];

    for (let line of initialLines) {
      while (stripAnsi(line).length > 0) {
        let counter = 0;
        let idx = 0;
        let inEscapeSequence = false;

        while (idx < line.length && counter < this.#TERMINAL_WIDTH) {
          if (line[idx] === "\x1b") {
            inEscapeSequence = true;
          }

          if (!inEscapeSequence) {
            counter++;
          }

          idx++;

          if (inEscapeSequence && line[idx] === "m") {
            inEscapeSequence = false;
            idx++;
          }
        }

        resultLines.push(line.slice(0, idx));
        line = line.slice(idx);
      }
    }

    return resultLines;
  }
  async #render(element: elementsJs.PipeComponents) {
    let string = "";

    string += renderToANSIString(element, this.#TERMINAL_WIDTH);

    const lines = this.#splitByNewlineOrLength(string);
    await this.#lockRender(async () => {
      for (const line of lines) {
        await this.#streamWrite(`${line}\n`);
      }
    });
  }
}
export const consoleRender = new ConsoleRender();
