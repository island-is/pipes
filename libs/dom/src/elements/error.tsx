import { span } from "./css/span.js";
import { renderRow } from "./row.js";
import { renderSubtitle } from "./subtitle.js";

import type { SpecifixJSX } from "./jsx.js";

export type IError = SpecifixJSX<"Error", null, string>;
export const Error = (props: Omit<IError, "type" | "children">, children: string): IError => {
  return {
    type: "Error",
    ...props,
    children,
  };
};

export const renderError = {
  ansi:
    (render: (child: any, width: number) => string) =>
    (component: IError, width: number): string => {
      const errorTitle = span(
        renderSubtitle.ansi(render)(
          {
            type: "Subtitle",
            children: "  ❌ ERROR",
          },
          (width - 1) / 2,
        ),
        {
          margin: {
            left: 1,
          },
          color: "red",
          textAlign: "center",
          padding: {
            left: 4,
            right: 4,
          },
        },
      );

      const values = renderRow.ansi(render)(
        {
          type: "Row",
          width: [26],
          children: [
            errorTitle,
            span(
              render(component.children, width),
              {
                width: width / 2 - 1,
                fontStyle: "bold",
              },
              width / 2 - 1,
            ) as any,
          ],
        },
        width,
      );
      return span(
        values,
        {
          width: width,
          border: {
            top: true,
            bottom: true,
          },
        },
        width,
      ).join("");
    },

  markdown:
    (render: (child: any) => string) =>
    (component: IError): string => {
      return `**Error:** ${render(component.children)}`;
    },
};
