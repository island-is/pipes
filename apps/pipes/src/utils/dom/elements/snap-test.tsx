import assert from "node:assert";
import { basename, dirname, join } from "node:path";
import { it } from "node:test";
import { fileURLToPath } from "node:url";

import { SnapshotState } from "jest-snapshot";

import render from "../../ink/render.js";

import type { ReactNode } from "react";

// eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
function toMatchSnapshot(actual: any, testFile: string, testTitle: string) {
  const root = join(fileURLToPath(dirname(testFile)), "__snapshots__");
  const fileName = `${basename(fileURLToPath(testFile))}.snap`;
  const snapshotFile = join(root, fileName);
  console.log(snapshotFile);
  const snapshotState = new SnapshotState(snapshotFile, {
    updateSnapshot: process.env.SNAPSHOT_UPDATE ? "all" : "new",
  } as any);

  const matcher = snapshotState.match({
    testName: `${testTitle}-${testFile}`,
    received: actual,
    isInline: false,
  });

  const result = matcher;
  snapshotState.save();

  return result;
}

// eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
export const snapTest = (actual: any, testFile: string, testTitle: string) => {
  const result = toMatchSnapshot(actual, testFile, testTitle);
  assert(result.pass, `${testTitle} ${basename(testFile)} - does not match snapshot`);
};

export const testJSX = async (jsx: ReactNode, name: string, filename: string): Promise<void> => {
  await it(name, async () => {
    const rendered = await render(jsx, true);
    const value = rendered.value();
    console.log(`${name}\n${value}`);
    snapTest(value, filename, name);
  });
};
