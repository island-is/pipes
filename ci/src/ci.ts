import { createPipe } from "@island.is/pipes-core";

import { buildContext } from "./build.js";
import { config } from "./config.js";
import { devBuildOrderContext } from "./dev-image-with-build-order.js";
import { devImageInstallContext } from "./dev-image.js";
import { githubTestContext } from "./github-test.js";
import { lintContext } from "./lint.js";
import { testContext } from "./test.js";
import { workspaceTestContext } from "./workspace-test.js";

await createPipe(() => {
  const tasks = [
    devImageInstallContext,
    devBuildOrderContext,
    workspaceTestContext,
    buildContext,
    lintContext,
    testContext,
  ];
  if (config.action === "Release") {
    return [...tasks];
  }
  if (config.action === "Test") {
    return [...tasks, githubTestContext];
  }
  throw new Error("Not defined");
});
