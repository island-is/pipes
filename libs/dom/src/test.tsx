import { consoleRender } from "./ci.js";
import { PipesDOM } from "./dom.js";
import { summaryRender } from "./markdown.js";

const dashboard2 = (
  <PipesDOM.Group title="Test">
    <PipesDOM.Mask values={["Dashboard"]} />
    <PipesDOM.Container>
      <PipesDOM.Highlight>Note: User feedback will be available tomorrow.</PipesDOM.Highlight>
      <PipesDOM.Dialog dialogType="error" message="There was an error processing your request. WOWOWOWOW!" />
      <PipesDOM.Title>Dashboard Overview</PipesDOM.Title>
      <PipesDOM.Subtitle>Metrics for August 2023</PipesDOM.Subtitle>
      <PipesDOM.Success>Data fetched successfully!</PipesDOM.Success>
      <PipesDOM.Timestamp time={"12.22.2020"} />
      <PipesDOM.Failure>Failed fetching data!</PipesDOM.Failure>
      <PipesDOM.Warning>There was a slight discrepancy in the financial data. Please review.</PipesDOM.Warning>
      <PipesDOM.Table>
        <PipesDOM.TableHeadings>
          <PipesDOM.TableCell>Metric</PipesDOM.TableCell>
          <PipesDOM.TableCell>Value</PipesDOM.TableCell>
          <PipesDOM.TableCell>Change since last month</PipesDOM.TableCell>
        </PipesDOM.TableHeadings>
        <PipesDOM.TableRow>
          <PipesDOM.TableCell>Users</PipesDOM.TableCell>
          <PipesDOM.TableCell>4,500</PipesDOM.TableCell>
          <PipesDOM.TableCell>+15%</PipesDOM.TableCell>
        </PipesDOM.TableRow>
        <PipesDOM.TableRow>
          <PipesDOM.TableCell>Revenue</PipesDOM.TableCell>
          <PipesDOM.TableCell>$12,000</PipesDOM.TableCell>
          <PipesDOM.TableCell>-2%</PipesDOM.TableCell>
        </PipesDOM.TableRow>
        <PipesDOM.TableRow>
          <PipesDOM.TableCell>Subscriptions</PipesDOM.TableCell>
          <PipesDOM.TableCell>2,000</PipesDOM.TableCell>
          <PipesDOM.TableCell>+8%</PipesDOM.TableCell>
        </PipesDOM.TableRow>
      </PipesDOM.Table>
      <PipesDOM.List>
        <PipesDOM.ListItem>Tasks completed: 75</PipesDOM.ListItem>
        <PipesDOM.ListItem>Pending tasks: 12</PipesDOM.ListItem>
        <PipesDOM.ListItem>Overdue tasks: 3</PipesDOM.ListItem>
      </PipesDOM.List>
      <PipesDOM.Error>Failed to fetch advertisement data. Please check the ad server.</PipesDOM.Error>
      <PipesDOM.Highlight>Note: User feedback will be available tomorrow.</PipesDOM.Highlight>
      <PipesDOM.Divider />
      <PipesDOM.Link href="https://www.mbl.is">mbl.is</PipesDOM.Link>
      <PipesDOM.Row>
        <PipesDOM.Badge>Admin</PipesDOM.Badge>
        <PipesDOM.Text value="Logged in as: JohnDoe" />
      </PipesDOM.Row>
    </PipesDOM.Container>
  </PipesDOM.Group>
);

const summary = (
  <PipesDOM.Container>
    <PipesDOM.Title>Wow</PipesDOM.Title>
    <PipesDOM.Table>
      <PipesDOM.TableHeadings>
        <PipesDOM.TableCell>Metric</PipesDOM.TableCell>
        <PipesDOM.TableCell>Value</PipesDOM.TableCell>
        <PipesDOM.TableCell>Change since last month</PipesDOM.TableCell>
      </PipesDOM.TableHeadings>
      <PipesDOM.TableRow>
        <PipesDOM.TableCell>Users</PipesDOM.TableCell>
        <PipesDOM.TableCell>4,500</PipesDOM.TableCell>
        <PipesDOM.TableCell>+15%</PipesDOM.TableCell>
      </PipesDOM.TableRow>
      <PipesDOM.TableRow>
        <PipesDOM.TableCell>Revenue</PipesDOM.TableCell>
        <PipesDOM.TableCell>$12,000</PipesDOM.TableCell>
        <PipesDOM.TableCell>-2%</PipesDOM.TableCell>
      </PipesDOM.TableRow>
      <PipesDOM.TableRow>
        <PipesDOM.TableCell>Subscriptions</PipesDOM.TableCell>
        <PipesDOM.TableCell>2,000</PipesDOM.TableCell>
        <PipesDOM.TableCell>+8%</PipesDOM.TableCell>
      </PipesDOM.TableRow>
    </PipesDOM.Table>
  </PipesDOM.Container>
);

// Now you can use this to render to the terminal:
consoleRender.mount(dashboard2);
await consoleRender.render();
summaryRender.mount(summary);
await summaryRender.render();
