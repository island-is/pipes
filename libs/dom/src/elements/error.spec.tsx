import { describe, it } from "node:test";

import { renderToANSIString } from "../ci.js";

import { span } from "./css/span.js";
import { Error, renderError } from "./error.js"; // Assuming error component file is named error.js
import { LOREM_IPSUM } from "./lorem.ipsum.spec.js";
import { snapTest } from "./snap-test.js";
const TERMINAL_WIDTH = process.stdout.columns;

describe("Render error", () => {
  describe("ansi", () => {
    it("render simple string error", () => {
      const element = Error({}, "Error message");
      const value = renderError.ansi(renderToANSIString)(element, process.stdout.columns);
      console.log(JSON.stringify(value));
      console.log(`\n${value}`);
      snapTest(value, import.meta.url, "render simple string error");
    });
  });
});
