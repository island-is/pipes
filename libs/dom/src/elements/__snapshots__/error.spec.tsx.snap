// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render colored error message-file:///var/home/lommi/Projects/island.is-pipes/libs/dom/src/elements/error.spec.tsx 1`] = `"                                                                                                                                                                                                                                    ────────────────────────────────────────────────────────────────────────────[4m[1m=== ❌ ERROR ===[22m[24m                        [1m[43m[31mColored Error: Something went wrong[0m[49m  [22m────────────────────────────────────────────────────────────────────────────                                                                                                                                                                                                                                    "`;

exports[`render multi-line colored error-file:///var/home/lommi/Projects/island.is-pipes/libs/dom/src/elements/error.spec.tsx 1`] = `"                                                                                                                                                                                                                                    ──────────────────────────────────────────────────────────────────────────────────────────────────────────────────[4m[1m=== ❌ ERROR ===[22m[24m                        [43m[31mLorem ipsum dolor sit amet, consectetur adipiscing elit.                   Vivamus lacinia odio vitae vestibulum. Donec in efficit                                                           ur leo. In hac habitasse platea dictumst. Proin dolor me                                                          tus, pharetra in ipsum a, facilisis posuere velit. Quisq                                                          ue sit amet eros eu urna dignissim pharetra non id liber                                                          o.                                                                                                                [0m[49m                                                                                                         ──────────────────────────────────────────────────────────────────────────────────────────────────────────────────                                                                                                                                                                                                                                    "`;

exports[`render multi-line error-file:///var/home/lommi/Projects/island.is-pipes/libs/dom/src/elements/error.spec.tsx 1`] = `"                                                                                                                                                                                                                                    ──────────────────────────────────────────────────────────────────────────────────────────────────────────────────[4m[1m=== ❌ ERROR ===[22m[24m                        Lorem ipsum dolor sit amet, consectetur adipiscing elit.                   Vivamus lacinia odio vitae vestibulum. Donec in efficit                                                           ur leo. In hac habitasse platea dictumst. Proin dolor me                                                          tus, pharetra in ipsum a, facilisis posuere velit. Quisq                                                          ue sit amet eros eu urna dignissim pharetra non id liber                                                          o.                                                                                                                                                                                                                                  Phasellus bibendum, odio nec tincidunt venenatis, velit                                                           orci bibendum augue, a molestie mi est eget tellus. Ut a                                                          uctor aliquet felis a cursus. Vestibulum ante ipsum prim                                                          is in faucibus orci luctus et ultrices posuere cubilia c                                                          urae; Sed feugiat, velit nec gravida interdum, enim nequ                                                          e fringilla mi, at volutpat orci dui at metus. Aenean vi                                                          verra auctor orci, eu mattis odio egestas a.                                                                                                                                                                                        Maecenas eleifend, sem eu lacinia cursus, orci risus sem                                                          per urna, in feugiat metus nulla ac est. Duis eget risus                                                          in nunc sollicitudin facilisis at nec magna. Nam vel ti                                                           ncidunt mauris. Pellentesque habitant morbi tristique se                                                          nectus et netus et malesuada fames ac turpis egestas. Vi                                                          vamus convallis, diam ac porta tempus, justo libero aliq                                                          uet augue, a eleifend libero elit eu nisi.                                                                        ──────────────────────────────────────────────────────────────────────────────────────────────────────────────────                                                                                                                                                                                                                                    "`;

exports[`render simple string error-file:///var/home/lommi/Projects/island.is-pipes/libs/dom/src/elements/error.spec.tsx 1`] = `"                                                                                                                                                                                                                                    ───────────────────────────────────────────────────────────────────────────────────────────────[4m[1m=== ❌ ERROR ===[22m[24m                        [1mError message                                           [22m───────────────────────────────────────────────────────────────────────────────────────────────                                                                                                                                                                                                                                    "`;
